src/Microsoft.AspNetCore.Routing/Matching/DefaultEndpointSelector.cs

```diff
@@ -24,16 +24,30 @@ public DefaultEndpointSelector(IEnumerable<MatcherPolicy> matcherPolicies)
             _selectorPolicies = matcherPolicies.OrderBy(p => p.Order).OfType<IEndpointSelectorPolicy>().ToArray();
         }
 
-        public override Task SelectAsync(
+        public override async Task SelectAsync(
             HttpContext httpContext,
             EndpointFeature feature,
             CandidateSet candidateSet)
         {
+            var selectorPolicies = _selectorPolicies;
             for (var i = 0; i < _selectorPolicies.Length; i++)
             {
-                _selectorPolicies[i].Apply(httpContext, candidateSet);
+                await selectorPolicies[i].ApplyAsync(httpContext, feature, candidateSet);
+                if (feature.Endpoint != null)
+                {
+                    // This is a short circuit, the selector chose an endpoint.
+                    return;
+                }
             }
 
+            ProcessFinalCandidates(httpContext, feature, candidateSet);
+        }
+
+        private static void ProcessFinalCandidates(
```

Comment:
FYI this isn't just for cleanliness, we're using ref var which requires us to make it non-async. We might need to to think about changes to this data structure if it's too hard to use in async methods.